dll/win32/shell32/CDefaultContextMenu.cpp

```diff
@@ -513,53 +539,96 @@ CDefaultContextMenu::AddStaticContextMenusToMenu(
             continue;
         }
 
-        BOOL Extended = FALSE;
+        UINT cmdFlags = 0;
+        BOOL hide = FALSE;
         HKEY hkVerb;
         if (idResource > 0)
         {
-            if (LoadStringW(shell32_hInstance, idResource, wszVerb, _countof(wszVerb)))
-                mii.dwTypeData = wszVerb; /* use translated verb */
-            else
-                ERR("Failed to load string\n");
-
-            LONG res = RegOpenKeyW(info.hkClass, wszKey, &hkVerb);
-            if (res == ERROR_SUCCESS)
+            if (!(uFlags & CMF_OPTIMIZEFORINVOKE))
             {
-                res = RegQueryValueExW(hkVerb, L"Extended", NULL, NULL, NULL, NULL);
-                Extended = (res == ERROR_SUCCESS);
-
-                RegCloseKey(hkVerb);
+                if (LoadStringW(shell32_hInstance, idResource, wszVerb, _countof(wszVerb)))
+                    mii.dwTypeData = wszVerb; /* use translated verb */
+                else
+                    ERR("Failed to load string\n");
             }
+
+            if (RegOpenKeyW(info.hkClass, wszKey, &hkVerb) != ERROR_SUCCESS)
+                hkVerb = NULL;
         }
         else
         {
-            LONG res = RegOpenKeyW(info.hkClass, wszKey, &hkVerb);
-            if (res == ERROR_SUCCESS)
+            if (RegOpenKeyW(info.hkClass, wszKey, &hkVerb) == ERROR_SUCCESS)
             {
-                DWORD cbVerb = sizeof(wszVerb);
-                res = RegLoadMUIStringW(hkVerb, NULL, wszVerb, cbVerb, NULL, 0, NULL);
-                if (res == ERROR_SUCCESS)
+                if (!(uFlags & CMF_OPTIMIZEFORINVOKE))
                 {
-                    /* use description for the menu entry */
-                    mii.dwTypeData = wszVerb;
-                }
+                    DWORD cbVerb = sizeof(wszVerb);
 
-                res = RegQueryValueExW(hkVerb, L"Extended", NULL, NULL, NULL, NULL);
-                Extended = (res == ERROR_SUCCESS);
+                    LONG res = RegLoadMUIStringW(hkVerb, L"MUIVerb", wszVerb, cbVerb, NULL, 0, NULL);
+                    if (res || !*wszVerb)
+                        res = RegLoadMUIStringW(hkVerb, NULL, wszVerb, cbVerb, NULL, 0, NULL);
 
-                RegCloseKey(hkVerb);
+                    if (res == ERROR_SUCCESS && *wszVerb)
+                    {
+                        /* use description for the menu entry */
+                        mii.dwTypeData = wszVerb;
+                    }
+                }
+            }
+            else
+            {
+                hkVerb = NULL;
             }
         }
 
-        if (!Extended || GetAsyncKeyState(VK_SHIFT) < 0)
+        if (hkVerb)
         {
-            mii.cch = wcslen(mii.dwTypeData);
+            // FIXME: GetAsyncKeyState should not be called here, clients
+            // need to be updated to set the CMF_EXTENDEDVERBS flag.
+            if (!(uFlags & CMF_EXTENDEDVERBS) && GetAsyncKeyState(VK_SHIFT) >= 0)
```

Comment:
In theory this should be `GetKeyState`, so you will get the keystate in sync with the processing, but if I recall correctly there should also be some keystate passed into the IContextMenu that should probably be used?

Reply:
It should not be GetKeyState either because the caller is supposed to supply CMF_EXTENDEDVERBS if shift is down. I will fix this in DefView and other places later but I already have a unrelated DefView PR that touches the menu slightly waiting...